### PR TITLE
Order by tuple

### LINKED ISSUE #38772: ORDER BY with columns in braces
Performance of ORDER BY on sorting key prefix with braces is 2x slower than without due to inefficient execution plan:
Without braces:
```sql
:) explain pipeline select CounterID, EventDate from hits_v1 order by CounterID, EventDate

EXPLAIN PIPELINE
SELECT
    CounterID,
    EventDate
FROM hits_v1
ORDER BY
    CounterID ASC,
    EventDate ASC

Query id: bc24ecf6-d364-4082-96d4-51dcfb591096

┌─explain───────────────────────────┐
│ (Expression)                      │
│ ExpressionTransform               │
│   (Sorting)                       │
│   MergingSortedTransform 19 → 1   │
│     (Expression)                  │
│     ExpressionTransform × 19      │
│       (ReadFromMergeTree)         │
│       MergeTreeInOrder × 19 0 → 1 │
└───────────────────────────────────┘
```
With braces:
```sql
:) explain pipeline select (a,b) from distinct_in_order order by (a,b)

EXPLAIN PIPELINE
SELECT (a, b)
FROM distinct_in_order
ORDER BY (a, b) ASC

┌─explain──────────────────────────┐
│ (Expression)                     │
│ ExpressionTransform              │
│   (Sorting)                      │
│   MergeSortingTransform          │
│     LimitsCheckingTransform      │
│       PartialSortingTransform    │
│         (Expression)             │
│         ExpressionTransform      │
│           (ReadFromMergeTree)    │
│           MergeTreeInOrder 0 → 1 │
└──────────────────────────────────┘
```

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

Fixes #38772

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
A less efficient execution plan can be generated for query with ORDER BY (a, b) than for ORDER BY a, b

### CODE REVIEW

**Comment by devcrafter** [comment]
@Mergifyio update

**Review by yakov-olkhovskiy — APPROVED** [review]
LGTM
consider maybe unnecessary changes

**Comment by yakov-olkhovskiy** [comment]
BTW this optimization makes WITH FILL works for such case 👍

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -886,25 +886,27 @@ static FillColumnDescription getWithFillDescription(const ASTOrderByElement & or
     return descr;
 }
 
-SortDescription InterpreterSelectQuery::getSortDescription(const ASTSelectQuery & query, ContextPtr context_)
+SortDescription InterpreterSelectQuery::getSortDescription(const ASTSelectQuery & query, const ContextPtr & context_)
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Interpreters/InterpreterSelectQuery.cpp:889]: Copying `ContextPtr` leads to counter increment, which is not necessary. There is a follow-up regarding this in #38902. I just touched the code in first commits of the fix

**Comment by devcrafter** [comment]
@Mergifyio update

File: tests/queries/0_stateless/02353_order_by_tuple.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+[ ! -z "$CLICKHOUSE_CLIENT_REDEFINED" ] && CLICKHOUSE_CLIENT=$CLICKHOUSE_CLIENT_REDEFINED
+
+##################
+# check that both queries have the same AST after rewrite, EXPLAIN SYNTAX returns it in form of query
+##################
+QUERY_ORDER_BY="SELECT number AS a, number % 2 AS b FROM numbers(10) ORDER BY a DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1, b DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1"
+QUERY_ORDER_BY_TUPLE="SELECT number AS a, number % 2 AS b FROM numbers(10) ORDER BY (a, b) DESC NULLS FIRST WITH FILL FROM 2 TO 1 STEP -1"
```
Comment (novikd) [inline tests/queries/0_stateless/02353_order_by_tuple.sh:13]: @devcrafter I'm not sure if we can do such an optimization. Is the query without optimization and with this `WITH FILL` valid?

**Pull request merged**